AppleALC/kern_alc.cpp

```diff
@@ -89,7 +89,7 @@ void AlcEnabler::updateProperties() {
 		// Secondly, update HDEF device and make it support digital audio
 		if (devInfo->audioBuiltinAnalog && validateInjection(devInfo->audioBuiltinAnalog)) {
 			const char *hdaGfx = nullptr;
-			if (hasBuiltinDigitalAudio && !devInfo->audioBuiltinDigital)
+			if (!devInfo->audioBuiltinDigital || !hasBuiltinDigitalAudio)
```

Comment:
This looks like junk to me. `onboard-X` properties are used to aggregate devices of digital audio. So you inject `onboard-X` into your IGPU device as well as into IGPU HDAU device. You do the same thing for GFX0 and its HDAU. Please remove or explain the choice.

Reply:
WIthout this change, when the digital audio is disabled, the analog audio properties are created with null in place of "hda-gfx". In such a case, the analog audio is disabled, and not working.

The new check is to test if digital audio is disabled, hence "onboard-1" is not being used so far this point. In such a case, it is using "onboard-1" when creating the analog audio. The digital audio can be disabled by two flags: devInfo->audioBuiltinDigital and hasBuiltinDigitalAudio.

Reply:
No, that logic is definitely flawed, even though I am not entirely sure where yet. Most likely we do not have enough understanding of the problem, but `No-hda-gfx` and `hda-gfx` simply cannot coexist in the same HDEF device, this is definitely borked.

`onboard-X` property, where X is a unique index, corresponds to pairs of digital devices on all native Macs. If you check IOReg dumps (e.g. from here https://applelife.ru/threads/dampy-originalnyx-makov.2943712/), you will see that the only case when Macs have `onboard-X` property in HDEF is when HDEF is both a digital audio device and an analog audio device. 

This is exactly what is present in the current code:
- `hasBuiltinDigitalAudio` (i.e. IGPU is active with a framebuffer that has connectors)
- `!devInfo->audioBuiltinDigital` (i.e. separate IGPU digital audio device, HDAU, is not used; i.e. we are not on Haswell, which is the only model that has HDAU device).

In all other cases on Macs HDEF device never has `onboard-X`, this is a rule. However, what is probably important is that there exist two things:
- there seems to be no Mac that has no device at all with `onboard-X` property. For all devices that have no `onboard-X` in HDEF, this property seems to exist in GPU device pairs (i.e. GFX0 and its HDAU).
- `No-hda-gfx` property is only present in iMac18,3 so far, and this Mac has an external GPU once again.

At the moment I have a Haswell setup with an AMD GPU. I have HDAU device entirely disabled, and my HDEF has no `onboard-X` property, it's only my GPU device pair (GFX0 and its HDAU) that have `onboard-1`. Obviously I have just fine working digital audio from my GPU and working analog audio from HDEF.

As a result I currently have a suspect that issue of not working analog audio comes with something like the following (hypothetical):
- you have Clover, which injected `hda-gfx` into IGPU or some other device on its own, and thus broke AppleHDA; to me this sounds very likely, as we cannot reproduce broken HDEF here with OpenCore. If you are not positive about it, consider reverting to master, injecting `No-hda-gfx`, and sending IOReg.
- we have no `hda-gfx` property at all, and macOS expects it to be there for some reason, theories follow:
   -  `No-hda-gfx` changes the rules, and once you have it you must have at least one `hda-gfx` somewhere.
   - Perhaps it is just a must to have `hda-gfx` when you have an IGPU with active connectors or something.

Reply:
Yeah, got your points. Thanks for all those info. For now, you can discard this pull request.